### DISCUSSION TITLE
Request for Comment: Cart Functionality

### DISCUSSION THREAD

**Discussion by morganmccunn** [Shopify/storefront-api-feedback#12] [RFC]
Traditionally, merchants using the Storefront API are using a checkout object as a cart. However, this has led to throttling issues for some merchants due to the higher level of protection we need to place on the checkout. That said, we set out to build a separate cart in the Storefront API that is not subject to the same throttling restrictions as checkout, and that provides price transparency for buyers in an effort to minimize surprises at checkout. 

The new cart endpoints in the Storefront API are currently in the unstable version so that we can get your input! Please take it for a spin and let us know what you think. 

We will do our best to address the feedback for this first version of the cart prior to v2021-10, and we do plan to continue building out functionality into the cart over time. For full details, check out the cart tutorial [here](https://shopify.dev/custom-storefronts/cart).

**Comment by dotcollective-brendon** [comment]
Excellent concept and this would resolve concerns for larger merchants, although this would need to support Shopify Scripts. Without that, there would still be a need to use the Checkout for cart sessions.

**Reply by morganmccunn** [reply]
Appreciate the feedback @b0123498765! We definitely understand the importance of scripts in the cart for price transparency. We’ll be working to incorporate scripts in the cart in the near future. Worth mentioning that scripts will still work at checkout for merchants who are keen to adopt the cart asap. We'll keep you posted once we've made progress on this.

**Comment by keidarcy** [comment]
I don't know is the correct place to say, but I think the cart `Ajax API` lacks the functionality to update multiple products based on line item property value. `/cart/change.js` supports update by property value, `/cart/update.js` supports update multiple items, but together I have to flush all cart information and add again. I know it's tricky but glad to use `storefront-api` if it provides this functionality.

**Reply by morganmccunn** [reply]
Hey @keidarcy - you can use the `cartLinesUpdate` mutation to update multiple line items. Curious if you have an example of what you're trying to experience for the buyer experience here. Let me know if you have additional details you can share, thanks!

**Comment by maddalax** [comment]
Great stuff, another benefit that this brings is the ability to add items with a selling plan for subscriptions. We've been limited by that for quite some time now and have had to do some hacky workarounds to make it work. I'll definitely be testing it out.

**Reply by morganmccunn** [reply]
Please do! And let us know if you run into any snags as you do @maddalax

**Reply by maddalax** [reply]
@morganmccunn Hi there, just wanted to make sure you saw this: https://github.com/Shopify/storefront-api-feedback/discussions/16

**Comment by razza12** [comment]
In the tutorial considerations it lists: "Shopify automatically deletes the cart when the customer completes their checkout."

By what mechanism is this meant to occur? After completing checkouts we are still able to query the cart ID - with no notification about the cart having been completed.

**Reply by morganmccunn** [reply]
Hey @razza12, curious what you would like a notification for? Right now, once the buyer enters the checkout flow they are redirected to the thank you page at the end of the flow (at which point the cart should be deleted).

**Reply by razza12** [reply]
@morganmccunn How is the cart deleted? After completing the checkout flow and reaching the thankyou page - the cart still seems to be available. If a user goes back to the frontend - the cart query (as at https://shopify.dev/custom-storefronts/cart#retrieve-a-cart) for the previous ID still returns the same response as before checking out. We even seem to be able to add additional items to the completed cart - although when checking out it goes straight back to the thank you page.

**Reply by cocoahero** [reply]
@razza12 This was a bug on our end that has now been fixed. You are correct that after a successful checkout, requesting the cart by ID will return `null`. We are looking into additional ways of communicating a successful checkout event beyond this behavior.

**Reply by EricMcRay** [reply]
@cocoahero I am still having this issue. After successful payment cart is still the same. It is working in the development store but not in the real store. I read that offsite payment gateways brokes the cart. Is it true? 

I checked abandoned checkouts and orders, they don't have `cart_token`.

**Reply by joepagan** [reply]
It appears that a completed order will **NOT** return `null` if a user checks out with a `Shop Pay` or `G Pay` the Express Checkout option (and probably others).
It sounds like this is still a bug.
How can we successfully determine this in headless apps otherwise?

**Reply by joepagan** [reply]
From Shopify support:

> After publishing [version 2023-01](https://shopify.dev/api/release-notes/2023-01) of Shopify's APIs, we've gone back and reviewed the remaining affected example you shared where you were receiving a non null response ("cart": {"id":"xxxxxxxxxxx"). We're seeing that after this latest update, the current supported version of the Storefront API receives null responses, and that the same request on depreciated versions, receives error responses.

So basically, express checkouts have never worked until the 2023-01 version!

**Comment by ahmed-adly-khalil** [comment]
Hello @morganmccunn 
One of the obstacles we have faced while working with the currently available cart API implementation is the lack of metadata. adding metadata to the new API like cart creation time, adding item per item, and some fields related to the user session would be very helpful in building functionality like cart recovery

**Reply by morganmccunn** [reply]
Hi @ahmed-adly-khalil  - apologies for the delayed reply! Cart recovery is definitely something we want to support and we're looking to understand what needs to be done for this. We are doing this in tandem of figuring out how we can better support analytics for the overall storefront api so this will take some time but I will report back when I have more information here!

**Reply by ahmed-adly-khalil** [reply]
@morganmccunn looking back into this after 1 year, i guess the web pixel extension is a good answer for analytics based scenarios like cart recovery? TY

**Comment by just-another-jxliu** [comment]
How are items ordered in the Cart? In my brief trial of the Cart API, it seems to have the same annoying behavior as the Checkout object where item order does not reflect the order in which the items were added by the customer. Any plans to change this behavior?

Are there any plans to make Cart objects accessible via the Admin API? That would allow our own abandon cart notification system to be more accurate.

**Reply by morganmccunn** [reply]
> Any plans to change this behavior?

Thanks for the feedback @just-another-jxliu - I understand why this might be preferred from a customer's perspective. Let me see what we might be able to do here in the future. 

> Are there any plans to make Cart objects accessible via the Admin API?

No not at the moment. Can you please elaborate on how this would impact your abandoned cart notification? Right now we know abandoned cart emails don't trigger from the SF API but I was leaning towards solving that and what you've suggested here. Curious to hear more about this though!

**Reply by haeky** [reply]
> How are items ordered in the Cart? In my brief trial of the Cart API, it seems to have the same annoying behavior as the Checkout object where item order does not reflect the order in which the items were added by the customer. Any plans to change this behavior?

@just-another-jxliu The merchandise lines should be ordered by their addition into the cart. Could you share your steps to reproduce what you're seeing?

**Comment by Grsmto** [comment]
It would be handy to be able to have the total number of items in a cart in the `CartLineConnection` object. 
Currently to show the number of items in a cart, I believe the only way is to fetch every cart items and add up the quantities of each line, which is not really efficient.

**Reply by morganmccunn** [reply]
Hmm fair suggestion @Grsmto ! Thanks for the feedback, let me get this on the list.

**Comment by m-sanders** [comment]
Are there any plans for the cart mutations to support capturing marketing attribution data? 

An issue that we've faced for a long time with our headless Checkout API is that there's no way (that I know of) to capture attribution data like UTMs that the online shop automatically captures. This breaks a lot of the reporting functionality in Shopify or at least limits its usefulness.

We currently persist attribution data to local storage and pass the data in as a JSON blob on checkout creation as a custom attributes entry. We also call the GraphQL endpoint with the attribution params when creating the checkout: `"\/api\/2021-07\/graphql.json?fbclid=...&utm_campaign=...&utm_medium=...&utm_source=...&utm_term=...`. We can then pull this data out in our own reporting. But we are not able to use any of the Shopify ecosystem for reporting channel attribution.

**Reply by morganmccunn** [reply]
Hey @m-sanders great question. Yes we absolutely intend to address this inconsistency between online store reporting and storefront api reporting, specifically with marketing attribution. Given some of the changes coming into play with regulations, there are changes occurring across the platform to ensure this is effectively supported. Once we establish this foundation we will be able to incorporate the changes into the storefront api cart.

**Reply by arobbins** [reply]
@morganmccunn @johncraigcole Any update on this?

**Comment by njbmartin** [comment]
Have raised a bug around international pricing and the cart functionality: https://github.com/Shopify/storefront-api-feedback/discussions/29. It is not being persisted through to checkout.

**Reply by morganmccunn** [reply]
Thank you! This is now fixed.

**Comment by benshrimpton** [comment]
When we saw the `cart` object added to the API, we got excited -- finally we can create carts, add discount codes, etc. etc. all on the front end of a Shopify theme..... **NOOOOOOPE!!**

**Reply by morganmccunn** [reply]
Hey @benshrimpton - you're correct in that the storefront API cart is separate from the cart currently used to power the Online Store (and thus used by themes) at the moment. However, I would be interested in learning more about the challenges you're having and what you were hoping to use a cart api for in more detail. Thanks!

**Reply by benshrimpton** [reply]
@morganmccunn it would be nice if you explicitly stated -- `this cart API is intended for headless applications as a way of bypassing the checkout API throttle rates and has no real advantage when used with a theme, if you want traditional Shopify cart features, don't use it, use the cart ajax API`. But you don't make that clear, so many hours later, we found that out. If you've not figured out what the differences are and why they would matter to developers, then I'm speechless. You're probably going to tell me to look on the Shopify App store now so you can make your 40% commission on a third party service...

**Reply by morganmccunn** [reply]
Okay, so it sounds like our documentation wasn't clear enough. I can understand that there is confusion between "storefront api" and the online store / themes platform. Thanks for the feedback - I will work with the team to see where we can be more explicit in the docs wrt the cart being for storefront API users only (which does not power the online store / themes platform)!  

I was asking for more explicit details in an effort to better understand your specific challenges. If you'd prefer not to share, then I can continue to rely on clear feedback from other developers who help us empathize with what they are trying to achieve, so we can plan our future efforts accordingly.

**Comment by patjakubik** [comment]
For us, creating the cart doesn't feel intuitive.

Currently, there are two options:
- `cartCreate` mutation, without items on first page visit
- `cartCreate` mutation with item, on first add to cart

It would be much easier to just call `cartLinesAdd` mutation without `cartId` and have cart created. However, currently `cartId` is required in `cartLinesAdd` mutation.

**Reply by nboliver** [reply]
Agree with this 💯 
There is a bunch of extra logc that's not required when the cart doesn't even exist yet.

**Reply by mrudult** [reply]
Also, the `cartLinesAdd` mutation returns the entire `Cart` object and should only return the added/updated item just like the Ajax `/cart/add.js` does

**Reply by jullite** [reply]
@mrudult 
I don't think so, the one added to cart may influence other products price in cart, so the entire Cart object need to be returned.

**Comment by apalumbo** [comment]
@morganmccunn There is an official date to Scripts support in Cart GRAPHQL Api?

**Reply by apalumbo** [reply]
@morganmccunn bump ;-)

**Reply by vixdug** [reply]
Hi @apalumbo, you can now utilize scripts with the Storefront API cart.

**Comment by lucadalli** [comment]
@morganmccunn @haeky
I have encountered two fairly disruptive bugs while working with the Cart API.
1. Cart line items are *not* returned in the order they were added. This was also an issue with the Checkout API and was [reported as early as 2018](https://github.com/Shopify/js-buy-sdk/issues/501). It baffles me that this is still unaddressed.
2. If two `cartLinesAdd` mutations run at the same time, the second mutation seems to overwrite the first, meaning that the `CartLineInput`s submitted with the first mutation are missing from the cart `lines`.

**Reply by vixdug** [reply]
Hi @lucadalli cart line items should now be returned in the order they were added to the cart.

**Comment by Asingerdev** [comment]
@morganmccunn have hit some roadblocks using the cart API. Would be really nice to be able to get shipping rates like you can from the checkout api, and some more metadata for user session. Also hit a bug with `cartLinesUpdate` where increasing the quantity of a variant with a selling plan converts all one-time variants to subscription, had to separate out one-time and subscription variants as a result. And an issue generally we're experiencing over and over with clients on custom storefronts - abandoned cart emails aren't automated. Not sure why this feature was removed, can reference this issue - https://github.com/Shopify/js-buy-sdk/issues/621

Thanks!

**Comment by evr-novicell** [comment]
Any news on applying Shopify Scripts on cart created/updated by Storefront API? I just had an "aha" moment looking for a solution and found out about storefront API's cart mutations. Then I found out it does not support Shopify Scripts 😬

**Reply by vixdug** [reply]
You can now utilize scripts with the Storefront API cart.

**Comment by Nortski78** [comment]
Can we have the ability to mutate the line item price? Essential for volume discounts.

**Reply by vixdug** [reply]
Hi @Nortski78 the price of a line item can not be mutated through the cart at this time. This can be achieved through [Shopify Scripts ](https://help.shopify.com/en/manual/checkout-settings/script-editor) or by creating a draft order.

**Reply by apalumbo** [reply]
@vixdug in the doc scripts are not supported in cart api, when this will be supported?

**Reply by vixdug** [reply]
Hi @apalumbo we've started to roll out scripts for the Cart API. I will update the thread when rollout is at 100%!

**Reply by apalumbo** [reply]
@vixdug Great is it possible to enable it as beta or anything similar?

**Reply by Nortski78** [reply]
> Hi @apalumbo we've started to roll out scripts for the Cart API. I will update the thread when rollout is at 100%!

Looking forward to this!

**Reply by vixdug** [reply]
Hey Everyone - Scripts is now 100% rolled out for the cart! If you have created your app more recently than the script itself, you may need to unpublish / republish the script 🥳 Thank you for your patience

**Comment by TylerP33** [comment]
How are we suppose to handle the persistence of a cart through user login/logouts?  You can easily access a customer with a customerAccessToken but accessing a cart on a re-login seems to be a tough task.  Naturally, when a user logs out, we clear out that token and cartId for security reasons.  When they log back in, they're given a new token but there doesn't seem to be a proper way to fetch their prior cart info.  Am I doing something wrong or is this going to require a workaround?

**Reply by joshistoast** [reply]
There's a `lastIncompleteCheckout` field on the `Customer` object that could be used, not sure if it returns a cart that hasn't been turned into a checkout or not.

**Reply by vixdug** [reply]
It would only return the abandoned checkout for that customer. Persistent cart for logged in buyers is something we are working on. I'd love to know what type of expectation you have around TTL of the saved cart and why.

**Reply by joshistoast** [reply]
@vixdug I think 14 days is a good TTL, if more time is needed then the customer is really just using the cart as a wishlist. Stores should circumvent this by providing a wishlist tool to customers to prevent this behavior.
